ci: run tests on windows and macos

legobeat opened the pull request:
This runs unit tests on `macOS-latest` and `windows-latest`, in addition to existing `ubuntu-latest`.

Only runs on node version from `.nvmrc` (currently v18).

#### Related

- Resolves #679
- WIP for actually making tests pass on Windows: https://github.com/legobeat/LavaMoat/pull/7

boneskull commented on .github/workflows/nodejs.yml:
```diff
@@ -32,3 +32,28 @@ jobs:
+      FORCE_COLOR: 1
+    steps:
+      - name: Checkout
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3
```
```suggestion
        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # tag=v3
```

for renovate

boneskull commented on .github/workflows/nodejs.yml:
```diff
@@ -32,3 +32,28 @@ jobs:
+      - name: Checkout
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 # v3
+      - name: Use Node.js
+        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3
```
```suggestion
        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # tag=v3
```

boneskull commented on .github/workflows/nodejs.yml:
```diff
@@ -32,3 +32,28 @@ jobs:
+      - name: Use Node.js
+        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3
+        with:
+          node-version-file: .nvmrc
```
If we're going to keep this as its own job, can we just use a static version here?  `latest` or `lts` or whatever the default is.  I don't really want to add a build dependency upon `.nvmrc`.  I have mixed feelings about that file. :P

boneskull commented on .github/workflows/nodejs.yml:
```diff
@@ -32,3 +32,28 @@ jobs:
+    name: Test platform compatibility
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
```
I guess I'd lean more towards just adding these to the main build matrix.  Why not?

boneskull replied:
[this is how you do it](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrixexclude)

basically, add `exclude` prop to `matrix`, and specify the OS/node-version pairs we don't want to run.

boneskull commented:
@legobeat I think we need to allow scripts for `nx` so it can grab/build its native module(s).

boneskull commented:
@legobeat looks like I fixed mac anyway. we expect windows to fail, of course

boneskull commented:
See #706 ; we don't need to worry about `nx` and its ilk after that gets merged.

boneskull commented:
[shx](https://npm.im/shx) might be helpful for this sort of thing if you want to avoid writing more custom scripts in js.

legobeat committed:
chore(ci): run unit tests on macOS-latest and windows-latest

legobeat committed:
chore(ci): disable failing win-compat tests

- disable windows tests in ci for:
  - aa
  - allow-scripts
  - core
  - tofu
  - yarn-plugin-allow-scripts
  - webpack
- nicer way to limit which packages should be windows-tested

legobeat committed:
test(browserify): increase ava timeout from 2m to 4m

boneskull commented on .github/workflows/nodejs.yml:
```diff
@@ -48,3 +48,63 @@ jobs:
+      fail-fast: false
+    env:
+      FORCE_COLOR: 1
+      TEST_WORKSPACES: >-
+        --workspace=packages/${{
+          join(
+            fromJson('
+              [
+                "lavapack",
+                "preinstall-always-fail"
+              ]'
+            ), ' --workspace=packages/'
+          )
+        }}
```
this needs a comment, whatever it is

boneskull replied:
good god what is that

boneskull commented on packages/browserify/package.json:
```diff
@@ -57,7 +57,7 @@
       "test/*.spec.js"
     ],
-    "timeout": "2m",
+    "timeout": "4m",
```
another way, probably not worth doing:

1. remove the `ava` config from `browserify/package.json`
2. add a `ava.config.mjs` which reproduces everything but contains a dynamic timeout based on platform
😄

boneskull reviewed (approved):
this is fine by me, but I'd rather see if we can combine the jobs for DRY reasons
